### ISSUE TITLE
Request for inclusion: mongo-c-driver

### ISSUE THREAD

**Issue by DEGoodmanWilson** [bincrafters/community#484]
### Description of Problem, Request, or Question

Would like to hand over mongo-c-driver conan wrapper to bincrafters
https://github.com/DEGoodmanWilson/conan-mongo-c-driver

### Package Details (Include if Applicable)
* Package Name/Version: **mongo-c-driver/1.9.4**
* Operating System: **Linux and OS X**
* Compiler+version: **gcc 4.9–7, clang 3.9–4.0*, Apple clang 7.3–10.0*
* Docker image: n/a

### Notes

I've been using this for a while, and it feels ready for others to use. The only caveat is that I don't have Windows builds going because TBH I don't know how to set those up.

I also have mongocxx wrappers which I would like to contribute, in [a separate issue](https://github.com/bincrafters/community/issues/485)

**Comment by DEGoodmanWilson** [comment]
Worth noting that this module also provides libbson. That might be a showstopper until I can separate them out.

**Comment by SSE4** [comment]
@DEGoodmanWilson 
repository was create
CI has been enabled
you should have an GitHub invitation in your e-mail

**Comment by DEGoodmanWilson** [comment]
Tests are passing on `testing/1.9.4`. I've updated the build files to not upload to my personal bintray and given the project something closer to a proper README. Waiting for `stable/1.9.4` to build. Once that is done, can we move the default branch to `stable/1.9.4`

**Comment by Croydon** [comment]
@DEGoodmanWilson Please use https://github.com/bincrafters/conan-readme-generator for the readmes

Also the env variables (upload repo, credentials..) should be set by us in the CI UIs not in the config files itself. Could you please check this out @SSE4?

**Comment by Croydon** [comment]
For the people searching:
https://github.com/bincrafters/conan-mongocxx
https://github.com/bincrafters/conan-mongo-c-driver

**Comment by SSE4** [comment]
yes, all environment variables have been set automatically, no need to touch them

**Comment by SSE4** [comment]
I've also changed default branches to the stable

**Comment by promgamer** [comment]
@SSE4 maybe you want to checkout this ones:
- https://github.com/bisect-pt/conan-mongo-cxx-driver/
- https://github.com/bisect-pt/conan-mongo-c-driver

**Comment by Croydon** [comment]
@promgamer Please feel free to open pull request when you have further improvements

**Comment by promgamer** [comment]
@Croydon I'll surely do it. I'll need some time to check the diffs.

**Comment by promgamer** [comment]
@SSE4 couldn't understand your question, can you clarify?

**Comment by SSE4** [comment]
there is `conan-mongcxx` project, and you linked `conan-mongo-cxx-driver`
are these two packages for the same project, and just have different naming, or are these two packages completely unrelated entities?

**Comment by promgamer** [comment]
two packages for the same project. I'll check the diff between yours and mine (which was based on @DEGoodmanWilson too), in order to see if I can improve something.

**Comment by SSE4** [comment]
okay, but which name is an official when? I prefer conan package matching the official name, to avoid confusion for consumers

**Comment by promgamer** [comment]
[mongo-c-driver](https://github.com/mongodb/mongo-c-driver) and [mongo-cxx-driver](https://github.com/mongodb/mongo-cxx-driver)

**Comment by DEGoodmanWilson** [comment]
Yes, `mongo-cxx-driver` is the correct official name. I originally named my repository and Conan project `conan-mongocxx`, which was a mistake I failed to rectify before requesting it be moved over. See here: https://github.com/bincrafters/community/issues/485#issuecomment-429912346

**Comment by SSE4** [comment]
@DEGoodmanWilson I've taken a deeper look, and your recipe uses configure to build, but @promgamer uses CMake and it builds successfully for Windows, as far as I can see. also, his recipe is much shorter  and clean. I suggest to merge @promgamer changes.

**Comment by promgamer** [comment]
@SSE4 @DEGoodmanWilson should I make a PR to bincrafters' repo or @DEGoodmanWilson repo? How do you want to do it?

**Comment by DEGoodmanWilson** [comment]
@promgamer to the bincrafters repo. I'm going to decommission mine.

**Comment by promgamer** [comment]
@DEGoodmanWilson and @SSE4 

[Conan-Mongo-C-Driver@1.11.0/testing](https://github.com/bincrafters/conan-mongo-c-driver/tree/testing/1.11.0) waiting for your feedback.

Builds here:
 - https://travis-ci.org/bincrafters/conan-mongo-c-driver/builds/442210529
 - https://ci.appveyor.com/project/BinCrafters/conan-mongo-c-driver/builds/19551113

**Comment by SSE4** [comment]
few minor things:
- `requires = 'zlib/[~=1.2]@conan/stable'` - don't use version ranges
- add `homepage` attribute
- rename `source_subfolder` -> `_source_subfolder`, `build_subfolder` -> `_build_subfolder`
- move `cmake.install()` to package method (take a look at `conan-templates` project for an example)
- don't use `collect_libs`, list libraries explicitly
- please add GCC 8, Clang 5 & 6 to the `travis.yml`

questions:
- why OpenSSL is Windows/Mac only, while on Linux it's not required, however, it links with `crypto` and `ssl`: 
```
        if tools.os_info.is_linux:
            self.cpp_info.libs.extend(["rt", "ssl", "crypto", "dl", "pthread"])
```
- what about `shared` option? also `fPIC`?

**Comment by promgamer** [comment]
- ~~requires = 'zlib/[~=1.2]@conan/stable' - don't use version ranges~~ **fixed**
- ~~add homepage attribute~~ **fixed**
- ~~rename source_subfolder -> _source_subfolder, build_subfolder -> _build_subfolder~~ **fixed**
- ~~move cmake.install() to package method (take a look at conan-templates project for an example)~~ **fixed**
- ~~please add GCC 8, Clang 5 & 6 to the travis.yml~~ **fixed**

> don't use collect_libs, list libraries explicitly

 why not? I had the idea this is more "bullet proof" than declaring manually. What is the added value?

> why OpenSSL is Windows/Mac only, while on Linux it's not required, however, it links with crypto and ssl

OpenSSL is required on [Linux builds, but not on mac/Windows](http://mongoc.org/libmongoc/current/installing.html).

> what about shared option? also fPIC?

fPIC is already there:
```
        if self.settings.os != 'Windows':
            cmake.definitions['CMAKE_POSITION_INDEPENDENT_CODE'] = True
```

I had some trouble with using a static mongo-c in the past, but I don't really remember why. I'll look at it again.

**Comment by SSE4** [comment]
really, not, `tools.collect_libs` is very dumb and known to cause issues. it just lists libraries in alpha-numeric order, which doesn't always match with correct linking order, so it may fail with linkers for which linking order is important (e.g. GNU LD).
for OpenSSL, then you don't need to list `ssl` and `crypto` in `cpp_info.libs` - conan will add all libraries from dependencies automatically.
for fPIC - most of recipes have `fPIC` as an option, I prefer to follow the same convention.

**Comment by Croydon** [comment]
>         if self.settings.os != 'Windows':
            cmake.definitions['CMAKE_POSITION_INDEPENDENT_CODE'] = True

This code is not required anymore for some Conan versions now. Please remove it. Conan is setting the CMake definition automatically when the fPIC option is true.
Please add a fPIC **option** with the default value true.

**Comment by DEGoodmanWilson** [comment]
@promgamer 

Some feedback:
- I'm not seeing the version ranges for zlib, for some reason.
- The patchfile is really confusing. It looks like a lot of it is just moving existing lines around—the necessary additions and subtractions are not very clear. I'd prefer to have a cleaner patchfile to make it easier to rework it for other versions of conan-c-driver. Is it possible for you to clean it up? Also, could you add a little bit of detail as to why it was necessary to patch that particular file in that particular way?

**Comment by promgamer** [comment]
Static on Linux is working. Having a hard time to make it work on Windows. If someone has a clue (scroll to the end of the log): https://ci.appveyor.com/project/BinCrafters/conan-mongo-c-driver/builds/19585507/job/g3g8xto0wbhb6gmv

**Comment by Croydon** [comment]
(@promgamer you can click on the line number to get an URL which points to these specific line)

**Comment by promgamer** [comment]
@Croydon edited above

**Comment by SSE4** [comment]
@promgamer 
try to add something like this into your `package_info` method:
```
if not self.options.shared:
    self.cpp_info.defines.append('MONGOC_STATIC=1')
```

**Comment by promgamer** [comment]
@SSE4 from their CMake, they do it in this [line](https://github.com/mongodb/mongo-c-driver/blob/master/src/libmongoc/CMakeLists.txt#L631). Do we need to add that definition again?

**Comment by SSE4** [comment]
yes, as you're adding definition to the project linking against mongo, but mongo's CMake only affects mongo's build

**Comment by promgamer** [comment]
Building is passing both for [Windows](https://ci.appveyor.com/project/bincrafters/conan-mongo-c-driver/history) and [Linux/Mac](https://travis-ci.org/bincrafters/conan-mongo-c-driver/branches)

**Comment by SSE4** [comment]
@promgamer @DEGoodmanWilson are you planning to extract `bson` to its own conan package?

**Comment by promgamer** [comment]
Maybe wait until someone asks for it?

**Comment by SSE4** [comment]
@promgamer sounds reasonable, I didn't find any requests in conan wishlist.
I am promoting package to stable now, thanks for the contribution!

**Comment by SSE4** [comment]
CI passed for stable, I consider this one done.
